### PR TITLE
Feature/project contents

### PR DESCRIPTION
**Pull request by darrylmorton-digicatapult**
- [x] members routes (GET, PUT)
- [x] integration tests
- [x] knex.js db operations, migration and seed file for test cleanup
- [x] openapi-express response validation 
- [x] helm chart updates
- [x] readme.md and db.md for docs

### CODE REVIEW

File: .github/workflows/apply-template.yml [hunk 1]
```diff
@@ -0,0 +1,73 @@
+name: Apply Template
```
Comment (mattdean-digicatapult) [inline .github/workflows/apply-template.yml:1]: This file shouldn't exist. Did the rename happen?

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,10 @@
+node_modules
+coverage
+.DS_Store
+.nyc_output
+.idea
+.env
+.vscode
+
+github_token
```
Comment (mattdean-digicatapult) [inline .gitignore:9]: `github_token`?

File: app/api-v1/api-doc.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+const apiDoc = {
+  openapi: '3.0.3',
+  info: {
+    title: 'IdentityService',
+    version: API_VERSION,
+  },
+  servers: [
+    {
+      url: `http://localhost:${PORT}/${API_MAJOR_VERSION}`,
```
Comment (mattdean-digicatapult) [inline app/api-v1/api-doc.js:11]: This probably shouldn't be localhost. Are we planning on using the openapi merge methodology we used in WASP to override this?

File: app/api-v1/services/apiService.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    const memberUpdated = await updateMemberAliasDb({ address, alias })
+    const result = memberUpdated.length ? memberUpdated[0] : {}
+
+    return { statusCode: 200, result }
+  } else {
+    const memberCreated = await createMemberAliasDb({ address, alias })
+    const result = memberCreated.length ? memberCreated[0] : {}
+
+    return { statusCode: 201, result }
+  }
```
Comment (mattdean-digicatapult) [inline app/api-v1/services/apiService.js:23]: I think the logic here is correct but perhaps confusing. I'd suggest breaking up the db query into two queries (one that matches `address` and one that matches `alias`).

File: app/api-v1/validators/responseValidators.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
+// const OpenAPIResponseValidator = require('openapi-response-validator').default
```
Comment (mattdean-digicatapult) [inline app/api-v1/validators/responseValidators.js:1]: Unused file?

File: app/util/appUtil.js [hunk 6]
```diff
@@ -0,0 +1,153 @@
+const { ApiPromise, WsProvider } = require('@polkadot/api')
```
Comment (mattdean-digicatapult) [inline app/util/appUtil.js:1]: This is a really random app util with multiple concerns. Consider breaking into a couple of different files related to the `api` and the `auth`

File: docker-compose.yaml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+      --unsafe-ws-external
+      --unsafe-rpc-external
+      --rpc-cors all"
+    ports:
+      - 30333:30333
+      - 9944:9944
+      - 9933:9933
+    restart: on-failure
+  postgres-data:
+    image: postgis/postgis:9.6-2.5-alpine
```
Comment (mattdean-digicatapult) [inline docker-compose.yaml:19]: Don't need `postgis` here. More modern straight `postgres` please

File: migrations/20220110125410_initial-table-definitions.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  await knex.schema.createTable('members', (def) => {
+    def.uuid('id').defaultTo(uuidGenerateV4())
+    def.string('address', 50).notNullable()
+    def.string('alias', 50).notNullable()
+    def.datetime('created_at').notNullable().default(now())
+    def.datetime('updated_at').notNullable().default(now())
+
+    def.primary(['id'])
+    def.unique(['address', 'alias'])
+    def.index(['alias'])
```
Comment (mattdean-digicatapult) [inline migrations/20220110125410_initial-table-definitions.js:21]: I don't think these are right. You probably want two unique indexes (one on `address` and one on `alias`)

**Review by mattdean-digicatapult — CHANGES_REQUESTED** [review]
Starter for 10

File: app/api-v1/validators/responseValidators.js (hunk 5 above)
Comment (darrylmorton-digicatapult) [reply app/api-v1/validators/responseValidators.js]: deleted

File: docker-compose.yaml (hunk 7 above)
Comment (darrylmorton-digicatapult) [reply docker-compose.yaml]: updated

File: migrations/20220110125410_initial-table-definitions.js (hunk 8 above)
Comment (darrylmorton-digicatapult) [reply migrations/20220110125410_initial-table-definitions.js]: done

File: app/api-v1/api-doc.js (hunk 3 above)
Comment (darrylmorton-digicatapult) [reply app/api-v1/api-doc.js:11]: Still TBD

File: .gitignore (hunk 2 above)
Comment (darrylmorton-digicatapult) [reply .gitignore:9]: left over from private repos, removed. will also remove from template repo

File: app/api-v1/services/apiService.js (hunk 4 above)
Comment (darrylmorton-digicatapult) [reply app/api-v1/services/apiService.js]: db query split as suggested and updated service logic

File: app/api-v1/api-doc.js (hunk 3 above)
Comment (mattdean-digicatapult) [reply app/api-v1/api-doc.js:11]: Yeah that's fine. For now we can leave it and make this decision down the line

File: helm/vitalam-identity-service/Chart.yaml [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+type: application
+maintainers:
+  - name: digicatapult
+    email: opensource@digicatapult.org.uk
+    url: www.digicatapult.org.uk
+dependencies:
+  - name: postgresql
+    version: '^10.12.2'
+    repository: https://charts.bitnami.com/bitnami
+    condition: postgresql.enabled
```
Comment (mattdean-digicatapult) [inline helm/vitalam-identity-service/Chart.yaml:15]: Odds are the `vitalam-node` should be a dependency here, but that should probably be a separate PR. Can you create an issue for that when this merges?

**Review by mattdean-digicatapult — CHANGES_REQUESTED** [review]
Last couple of things. Really close. I'd like @dblane-digicatapult to review the helm chart also

File: helm/vitalam-identity-service/templates/configmap.yaml [hunk 10]
```diff
@@ -0,0 +1,12 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ include "vitalam-identity-service.fullname" . }}-config
+  labels:
+    {{- include "vitalam-identity-service.labels" . | nindent 4 }}
+data:
+  port: {{ .Values.config.port | quote }}
+  logLevel: {{ .Values.config.logLevel }}
+  dbHost: {{ include "vitalam-identity-service.fullname" . }}
```
Comment (dblane-digicatapult) [inline helm/vitalam-identity-service/templates/configmap.yaml:10]: `vitalam-identity-service.postgresql.fullname`

File: app/env.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    AUTH_ISSUER: envalid.url(),
+    AUTH_TOKEN_URL: envalid.url(),
+    METADATA_KEY_LENGTH: envalid.num({ default: 32 }),
+    METADATA_VALUE_LITERAL_LENGTH: envalid.num({ default: 32 }),
+    LOG_LEVEL: envalid.str({ default: 'info', devDefault: 'debug' }),
+    DB_HOST: envalid.host({ devDefault: 'localhost' }),
+    DB_PORT: envalid.port({ default: 5432 }),
+    DB_NAME: envalid.str({ default: 'vitalam' }),
+    DB_USERNAME: envalid.str({ devDefault: 'postgres' }),
+    DB_PASSWORD: envalid.str({ devDefault: 'postgres' }),
```
Comment (dblane-digicatapult) [inline app/env.js:32]: This needs sorting out, if we don't have `defaults` present here then they need to be added to `helm/values.yaml` and then correctly inserted into the corresponding templates.

File: helm/vitalam-identity-service/templates/configmap.yaml [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+  dbPort: {{ .Values.config.dbPort | quote }}
+  dbName: {{ .Values.config.dbName }}
```
Comment (dblane-digicatapult) [inline helm/vitalam-identity-service/templates/configmap.yaml:12]: We need to be have some of the `ENVARS` able to be put into this configmap

File: helm/vitalam-identity-service/templates/deployment.yaml [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+          image: {{ .Values.image.repository }}:{{ .Values.image.tag }}
+          imagePullPolicy: {{ .Values.image.pullPolicy | quote }}
+          livenessProbe:
+            httpGet:
+              path: /health
+              port: {{ .Values.config.port }}
+            initialDelaySeconds: 15
+            timeoutSeconds: 3
+            periodSeconds: 5
+          env:
```
Comment (dblane-digicatapult) [inline helm/vitalam-identity-service/templates/deployment.yaml:63]: Add missing `ENVARS` here from the `configmap`

File: helm/vitalam-identity-service/values.yaml [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+  enabled: true
+  postgresqlDatabase: vitalam
+  postgresqlUsername: postgres
+  postgresqlPassword: postgres
+config:
+  port: 80
+  logLevel: info
+  dbName: vitalam
+  dbPort: 5432
+  externalPostgresql: null
```
Comment (dblane-digicatapult) [inline helm/vitalam-identity-service/values.yaml:11]: Add defaults here if we are not setting them as defaults in `env.js`

**Review by dblane-digicatapult — CHANGES_REQUESTED** [review]
I think this is also going to have a dependency on the vitalam-node.  I guess we need to decide if this is only essential for chart testing and how we want to handle it in production. @mattdean-digicatapult

File: test/mocharc.js [hunk 15]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+  recursive: true,
+  slow: 2000,
+  timeout: 10000,
```
[the remark this answers is not in the document]
Comment (darrylmorton-digicatapult) [reply test/mocharc.js]: now using defaults

File: docs/db.md [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+| `created_at`    | `Timestamp with timezone` | FALSE    |       `now()`        | When the row was first created      |
+| `updated_at`    | `Timestamp with timezone` | FALSE    |       `now()`        | When the row was last updated       |
+
+#### Indexes
+
+| columns            | Index Type | description                                                                         |
+|:-------------------|:-----------|:------------------------------------------------------------------------------------|
+| `id`               | PRIMARY    | Primary key                                                                         |
+| `alias`            | Index      | Allows quick filtering of `members` by `alias` name                                 |
+| `address`, `alias` | Unique     | Prevents more than one `member` from being created with identical characterisations |
```
[the remark this answers is not in the document]
Comment (mattdean-digicatapult) [reply docs/db.md]: Have a look at your description strings @darrylmorton-digicatapult 😄

File: app/env.js (hunk 11 above)
Comment (mattdean-digicatapult) [reply app/env.js:32]: Unfortunately some of these just can't have sane defaults. The auth stuff in particular I don't know what we would put

File: helm/vitalam-identity-service/templates/configmap.yaml (hunk 12 above)
Comment (mattdean-digicatapult) [reply helm/vitalam-identity-service/templates/configmap.yaml:12]: Agreed by default all the envs should be configurable by the helm chart

**Comment by mattdean-digicatapult** [comment]
> I think this is also going to have a dependency on the vitalam-node. I guess we need to decide if this is only essential for chart testing and how we want to handle it in production. @mattdean-digicatapult

@dblane-digicatapult yes this came up in my review also. My temptation is to let this merge a bit broken form the helm front and then fix up the chart. I'm worried this PR is just getting too huge right now and given the service isn't being used anywhere seems like a low risk

**Comment by dblane-digicatapult** [comment]
> @dblane-digicatapult yes this came up in my review also. My temptation is to let this merge a bit broken form the helm front and then fix up the chart. I'm worried this PR is just getting too huge right now and given the service isn't being used anywhere seems like a low risk

I'd agree.  Creating a ticket in the backlog for this to be worked on later.

File: app/env.js (hunk 11 above)
Comment (darrylmorton-digicatapult) [reply app/env.js]: I've set the `auth` related envs. The application's `host` seems to be the exception and is not set in other helm charts I've looked around. Is this just not needed because of implicit defaults or missed?

File: helm/vitalam-identity-service/templates/configmap.yaml (hunk 12 above)
Comment (darrylmorton-digicatapult) [reply helm/vitalam-identity-service/templates/configmap.yaml]: done

File: helm/vitalam-identity-service/templates/deployment.yaml (hunk 13 above)
Comment (darrylmorton-digicatapult) [reply helm/vitalam-identity-service/templates/deployment.yaml:62]: done

File: helm/vitalam-identity-service/values.yaml (hunk 14 above)
Comment (darrylmorton-digicatapult) [reply helm/vitalam-identity-service/values.yaml:11]: done

File: helm/vitalam-identity-service/templates/configmap.yaml (hunk 10 above)
Comment (darrylmorton-digicatapult) [reply helm/vitalam-identity-service/templates/configmap.yaml]: done

File: docs/db.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (darrylmorton-digicatapult) [reply docs/db.md]: updated

File: .github/workflows/apply-template.yml [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+          git commit -m "Apply templating to ${{ needs.repo_ids.outputs.repo_name }}"
+      - name: Create Pull Request
+        uses: peter-evans/create-pull-request@v3
+        with:
+          branch: bot/apply-template
+          base: main
+          title: Apply template
+          body: Automated pull request to apply repository template changes
+          assignees: ${{ github.actor }}
+          reviewers: ${{ github.actor }}
```
Comment (dblane-digicatapult) [inline .github/workflows/apply-template.yml]: Remove this file

**Review by dblane-digicatapult — CHANGES_REQUESTED** [review]

File: .github/workflows/apply-template.yml (hunk 17 above)
Comment (darrylmorton-digicatapult) [reply .github/workflows/apply-template.yml]: done

**Review by mattdean-digicatapult — APPROVED** [review]

**Review by dblane-digicatapult — APPROVED** [review]
